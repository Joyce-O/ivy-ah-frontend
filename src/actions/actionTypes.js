export const AUTH_LOADING = 'AUTH_LOADING';
export const LOGIN = 'LOGIN';
export const TOGGLE_MODAL = 'TOGGLE_MODAL';
export const FETCH_ARTICLE_STARTED = 'FETCH_ARTICLE_STARTED';
export const FETCH_ARTICLE_SUCCEEDED = 'FETCH_ARTICLE_SUCCEEDED';
export const FETCH_ARTICLE_FAILED = 'FETCH_ARTICLE_FAILED';
export const EDIT_BOOKMARKS_STARTED = 'EDIT_BOOKMARKS_STARTED';
export const FETCH_BOOKMARKS_SUCCEEDED = 'FETCH_BOOKMARKS_SUCCEEDED';
export const ADD_BOOKMARKS_SUCCEEDED = 'ADD_BOOKMARKS_SUCCEEDED';
export const REMOVE_BOOKMARKS_SUCCEEDED = 'REMOVE_BOOKMARKS_SUCCEEDED';
export const EDIT_BOOKMARKS_FAILED = 'EDIT_BOOKMARK_FAILED';
export const AUTH_SUCCESS = 'AUTH_SUCCESS';
export const AUTH_FAIL = 'AUTH_FAIL';
export const VERIFY_EMAIL = 'VERIFY_EMAIL';
