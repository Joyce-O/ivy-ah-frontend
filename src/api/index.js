import axios from 'axios';

const API_BASE_URL = 'https://ivy-ah-backend-staging.herokuapp.com/api/v1';

const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6Iml2eWFkbWluQGdtYWlsLmNvbSIsImlkIjoiMDU1MDk2YWUtMDcyYy00MjJkLTgwMTctZDM2ZTFiZWNmNTNjIiwicm9sZSI6ImFkbWluIiwiaWF0IjoxNTU3MTUzNzc2fQ.SeUSBXKom7eHkEXgak8uS-3xqVKSwR3DvrGyIcn7CSs';

export const client = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
    Authorization: token
  },
});
export const signup = params => client.post('/users/signup', params);
export const login = params => client.post('/users/login', params);
export const fetchBookmarks = () => client.get('/bookmarks');
export const addBookmark = article => client.post('/bookmarks', { article });
export const removeBookmark = article => client.delete('/bookmarks', {
  data: {
    article
  }
});
export const fetchArticle = id => client.get(`/articles/${id}`);
