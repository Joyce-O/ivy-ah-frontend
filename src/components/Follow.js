import React from 'react';
import PropTypes from 'prop-types';

const FollowSwitch = ({
  onClickFollow, isFollowing, userId, authorId
}) => (
  userId !== authorId
    && (
      <div>
        <button type="button" className="ui button follow" onClick={() => onClickFollow(isFollowing)}>{isFollowing ? 'Unfollow' : 'Follow'}</button>
      </div>
    )
);

FollowSwitch.propTypes = {
  onClickFollow: PropTypes.func.isRequired,
  isFollowing: PropTypes.bool.isRequired,
  userId: PropTypes.string.isRequired,
  authorId: PropTypes.string.isRequired
};

export default FollowSwitch;
