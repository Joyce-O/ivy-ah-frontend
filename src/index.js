import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import 'semantic-ui-css/semantic.min.css';
import 'overlayscrollbars/css/OverlayScrollbars.min.css';
import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import App from './containers/App';
import Store from './store';
import './assets/sass/index.scss';

render(<Provider store={Store}><App /></Provider>, document.getElementById('root'));
