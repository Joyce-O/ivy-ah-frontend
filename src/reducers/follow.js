import * as actions from '../actions/actionTypes';

const initialFollowState = {
  following: [],
  isFollowing: false,
  error: null,
};

export default (state = initialFollowState, action) => {
  switch (action.type) {
    case actions.FETCH_FOLLOWING_SUCCEEDED: {
      return {
        ...state,
        following: action.payload.user.following,
      };
    }
    case actions.FOLLOW_USER_SUCCEEDED: {
      return {
        ...state,
        following: state.following.concat(action.payload.authorId),
        isFollowing: true
      };
    }
    case actions.UNFOLLOW_USER_SUCCEEDED: {
      return {
        ...state,
        following: state.following.filter(follows => follows.id !== action.payload.authorId),
        isFollowing: false
      };
    }
    default:
      return state;
  }
};
