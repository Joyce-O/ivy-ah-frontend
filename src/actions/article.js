import * as api from '../api';
import * as actions from './actionTypes';

export const fetchArticlesStarted = () => ({
  type: actions.FETCH_ALL_ARTICLES_STARTED,
});

const fetchArticlesSucceeded = articles => ({
  type: actions.FETCH_ALL_ARTICLES_SUCCEEDED,
  payload: { articles }
});

const fetchArticlesFailed = err => ({
  type: actions.FETCH_ALL_ARTICLES_FAILED,
  payload: { err }
});

export const fetchArticles = (page, limit) => (dispatch) => {
  dispatch(fetchArticlesStarted());
  return api.fetchArticles(page, limit).then((res) => {
    dispatch(fetchArticlesSucceeded(res.data));
  }).catch((err) => {
    dispatch(fetchArticlesFailed(err.message));
  });
};
