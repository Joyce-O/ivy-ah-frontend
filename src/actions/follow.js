import * as api from '../api';
import * as actions from './actionTypes';

export const followUserStarted = authorId => ({
  type: actions.FOLLOW_USER_STARTED,
  payload: { authorId }
});

export const followUserSucceeded = authorId => ({
  type: actions.FOLLOW_USER_SUCCEEDED,
  payload: { authorId }
});

export const followUser = authorId => (dispatch) => {
  dispatch(followUserStarted(authorId));
  return api.followUser(authorId).then((res) => {
    dispatch(followUserSucceeded(authorId));
  });
};

export const fetchFollowingStarted = userId => ({
  type: actions.FETCH_FOLLOWING_STARTED,
  payload: { userId }
});

export const fetchFollowingSucceeded = user => ({
  type: actions.FETCH_FOLLOWING_SUCCEEDED,
  payload: { user }
});

export const userFollowing = userId => (dispatch) => {
  dispatch(fetchFollowingStarted(userId));
  return api.fetchUserFollowing(userId).then((res) => {
    dispatch(fetchFollowingSucceeded(res.data.data[0]));
  });
};

export const unFollowUserStarted = authorId => ({
  type: actions.UNFOLLOW_USER_STARTED,
  payload: { authorId }
});

export const unFollowUserSucceeded = authorId => ({
  type: actions.UNFOLLOW_USER_SUCCEEDED,
  payload: { authorId }
});

export const unFollowUser = authorId => (dispatch) => {
  dispatch(unFollowUserStarted(authorId));
  return api.unFollowUser(authorId).then((res) => {
    dispatch(unFollowUserSucceeded(authorId));
  });
};
