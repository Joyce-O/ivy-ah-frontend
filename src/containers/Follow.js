import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { toggleModal } from '../actions/auth';
import Follow from '../components/Follow';
import { followUser, userFollowing, unFollowUser } from '../actions/follow';

export const FollowContainer = ({
  dispatch, userId, isFollowing, authorId
}) => {
  useEffect(() => {
    dispatch(userFollowing(userId));
  }, [userId]);

  const onClickFollow = (followed) => {
    if (!userId) {
      return dispatch(toggleModal('sign in'));
    }
    return followed ? dispatch(unFollowUser(authorId)) : dispatch(followUser(authorId));
  };


  return (
    <Follow
      onClickFollow={onClickFollow}
      isFollowing={isFollowing}
      userId={userId}
      authorId={authorId}
    />
  );
};

FollowContainer.propTypes = {
  dispatch: PropTypes.func.isRequired,
  isFollowing: PropTypes.bool.isRequired,
  userId: PropTypes.string.isRequired,
  authorId: PropTypes.string.isRequired
};

const mapStateToProps = (state, { authorId }) => {
  const { following } = state.follow;
  const { userId } = state.auth;
  const followingTheseAuthors = following.map(user => user.id);
  let { isFollowing } = state.follow;

  if (followingTheseAuthors.includes(authorId)) {
    isFollowing = true;
  }

  return {
    authorId,
    isFollowing,
    following: followingTheseAuthors,
    userId
  };
};

export default connect(mapStateToProps)(FollowContainer);
