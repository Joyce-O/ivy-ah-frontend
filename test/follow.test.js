import React from 'react';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import MockAdapter from 'axios-mock-adapter';
import * as api from '../src/api';
import * as actions from '../src/actions/actionTypes';
import {
  fetchFollowingStarted, userFollowing, followUserStarted,
  followUser, unFollowUserStarted, unFollowUser
} from '../src/actions/follow';
import follow from '../src/reducers/follow';
import FollowSwitch from '../src/components/Follow';
import { FollowContainer } from '../src/containers/Follow';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const mock = new MockAdapter(api.client);

describe('Tests for the Follow/Unfollow user action creators', () => {
  const userId = 'f60cd301-f922-4f02-8316-2012df279636';
  const authorId = '5dc35524-b257-496e-9264-a089db25ff4f';

  it('fetchFollowingStarted', () => {
    const expectedAction = {
      type: actions.FETCH_FOLLOWING_STARTED,
      payload: {
        userId
      }
    };

    expect(fetchFollowingStarted(userId)).toEqual(expectedAction);
  });

  it('fetchFollowingSucceeded', () => {
    let user;
    mock.onGet(`/profiles/${userId}/following`).reply(200, {
      data: {}
    });

    const getState = {};
    const expectedActions = [
      { type: actions.FETCH_FOLLOWING_STARTED, payload: { userId } },
      { type: actions.FETCH_FOLLOWING_SUCCEEDED, payload: { user } }
    ];

    const store = mockStore(getState, expectedActions);
    return store.dispatch(userFollowing(userId)).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('followUserStarted', () => {
    const expectedAction = {
      type: actions.FOLLOW_USER_STARTED,
      payload: { authorId }
    };

    expect(followUserStarted(authorId)).toEqual(expectedAction);
  });

  it('followUserSucceeded', () => {
    mock.onPost(`/profiles/${authorId}/follow`).reply(200, {
      data: {}
    });

    const getState = {};
    const expectedActions = [
      { type: actions.FOLLOW_USER_STARTED, payload: { authorId } },
      { type: actions.FOLLOW_USER_SUCCEEDED, payload: { authorId } }
    ];

    const store = mockStore(getState, expectedActions);
    return store.dispatch(followUser(authorId)).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('followUserStarted', () => {
    const expectedAction = {
      type: actions.FOLLOW_USER_STARTED,
      payload: { authorId }
    };

    expect(followUserStarted(authorId)).toEqual(expectedAction);
  });

  it('unFollowUserStarted', () => {
    const expectedAction = {
      type: actions.UNFOLLOW_USER_STARTED,
      payload: { authorId }
    };

    expect(unFollowUserStarted(authorId)).toEqual(expectedAction);
  });

  it('unFollowUserSucceeded', () => {
    mock.onDelete(`/profiles/${authorId}/follow`).reply(200, {
      data: {}
    });

    const getState = {};
    const expectedActions = [
      { type: actions.UNFOLLOW_USER_STARTED, payload: { authorId } },
      { type: actions.UNFOLLOW_USER_SUCCEEDED, payload: { authorId } }
    ];

    const store = mockStore(getState, expectedActions);
    return store.dispatch(unFollowUser(authorId)).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});

describe('Tests for the Follow Reducer', () => {
  const initialState = {
    following: [],
    isFollowing: false,
    error: null,
  };

  it('should handle initial state', () => {
    expect(follow(undefined, {})).toEqual(initialState);
  });

  it('should handle fetch following succeeded', () => {
    const action = {
      type: actions.FETCH_FOLLOWING_SUCCEEDED,
      payload: {
        user: {
          following: ['Array of authors the user is folloing']
        }
      }
    };
    const expectedState = {
      ...initialState,
      following: action.payload.user.following,
      isFollowing: false
    };

    expect(follow(initialState, action)).toEqual(expectedState);
  });

  it('should handle follow user succeeded', () => {
    const action = {
      type: actions.FOLLOW_USER_SUCCEEDED,
      payload: {
        authorId: '5dc35524-b257-496e-9264-a089db25ff4f'
      }
    };
    const expectedState = {
      ...initialState,
      following: initialState.following.concat(action.payload.authorId),
      isFollowing: true
    };

    expect(follow(initialState, action)).toEqual(expectedState);
  });

  it('should handle unfollow user succeeded', () => {
    const action = {
      type: actions.UNFOLLOW_USER_SUCCEEDED,
      payload: {
        authorId: '5dc35524-b257-496e-9264-a089db25ff4f'
      }
    };
    const expectedState = {
      ...initialState,
      following: initialState.following.filter(follows => follows.id !== action.payload.authorId),
      isFollowing: false
    };

    expect(follow(initialState, action)).toEqual(expectedState);
  });
});

describe('Tests for the Follow Component', () => {
  it('renders the button with Unfollow when isFollowing is true', () => {
    const userId = 'f60cd301-f922-4f02-8316-2012df279636';
    const authorId = '5dc35524-b257-496e-9264-a089db25ff4f';
    const onClickFollow = jest.fn();
    const isFollowing = true;
    const tree = renderer
      .create(<FollowSwitch
        onClickFollow={onClickFollow}
        isFollowing={isFollowing}
        userId={userId}
        authorId={authorId}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the button with Follow when isFollowing is false', () => {
    const userId = 'f60cd301-f922-4f02-8316-2012df279636';
    const authorId = '5dc35524-b257-496e-9264-a089db25ff4f';
    const onClickFollow = jest.fn();
    const isFollowing = false;
    const tree = renderer
      .create(<FollowSwitch
        onClickFollow={onClickFollow}
        isFollowing={isFollowing}
        userId={userId}
        authorId={authorId}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders <Follow Container />', () => {
    const wrapper = shallow(<FollowContainer
      dispatch
      userId
      authorId
      isFollowing
    />);

    expect(wrapper.find(FollowContainer).length).toEqual(0);
  });
});
