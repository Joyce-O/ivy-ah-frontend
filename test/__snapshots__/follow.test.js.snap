// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the Follow Component renders the button with Follow when isFollowing is false 1`] = `
<div>
  <button
    className="ui button follow"
    onClick={[Function]}
    type="button"
  >
    Follow
  </button>
</div>
`;

exports[`Tests for the Follow Component renders the button with Unfollow when isFollowing is true 1`] = `
<div>
  <button
    className="ui button follow"
    onClick={[Function]}
    type="button"
  >
    Unfollow
  </button>
</div>
`;
